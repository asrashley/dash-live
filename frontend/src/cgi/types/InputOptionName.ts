export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}