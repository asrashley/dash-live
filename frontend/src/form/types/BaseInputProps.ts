import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

