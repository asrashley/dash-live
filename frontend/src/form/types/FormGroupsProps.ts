import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
