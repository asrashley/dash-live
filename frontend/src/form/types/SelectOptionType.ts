export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
