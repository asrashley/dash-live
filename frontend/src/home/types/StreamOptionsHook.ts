import { createContext } from "preact";
import { UseStreamOptionsHook } from "../hooks/useStreamOptions";

export const StreamOptionsContext = createContext<UseStreamOptionsHook>(null);
