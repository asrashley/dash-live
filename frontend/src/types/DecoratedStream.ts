import { Stream } from "./Stream";
import { StreamTrack } from "./StreamTrack";


export interface DecoratedStream extends Stream {
    tracks: StreamTrack[];
}
