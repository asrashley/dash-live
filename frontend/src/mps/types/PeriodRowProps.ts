import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
