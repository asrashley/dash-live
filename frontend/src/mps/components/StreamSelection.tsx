import { useCallback, useContext } from "preact/hooks";

import { AllStreamsContext } from '../../hooks/useAllStreams';
import { DecoratedStream } from "../../types/DecoratedStream";

export interface StreamSelectionProps {
  value?: DecoratedStream,
  onChange: (props: {name: string, value: number}) => void;
  name: string;
  error?: string;
  required?: boolean;
}

export function StreamSelection({ value, onChange, name, error, required }: StreamSelectionProps) {
  const { allStreams } = useContext(AllStreamsContext);
  const validationClass = error
    ? " is-invalid"
    : value
    ? " is-valid"
    : "";
  const className = `form-select${validationClass}`;

  const changeHandler = useCallback(
    (ev) => {
      onChange({
        name,
        value: parseInt(ev.target.value, 10),
      });
    },
    [onChange, name]
  );

  return <select
    className={className}
    value={value?.pk}
    name={name}
    onChange={changeHandler}
    required={required}
  >
    <option value="">--Select a stream--</option>
    {allStreams.value.map((s) => <option key={s.pk} value={s.pk}>{s.title}</option>)}
  </select>;
}

