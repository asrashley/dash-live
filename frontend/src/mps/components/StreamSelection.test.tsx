import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import {
  AllStreamsContext,
  decorateAllStreams,
  UseAllStreamsHook,
} from "../../hooks/useAllStreams";
import { DecoratedStream } from "../../types/DecoratedStream";
import { StreamSelection } from "./StreamSelection";

import { streams } from "../../test/fixtures/streams.json";

describe("StreamSelection component", () => {
  const allStreams = signal<DecoratedStream[]>([]);
  const loaded = signal<boolean>(false);
  const streamsMap = signal<Map<string, DecoratedStream>>(new Map());
  const error = signal<string | null>(null);
  const useAllStreamsHookMock: UseAllStreamsHook = {
    allStreams,
    loaded,
    streamsMap,
    error,
  };
  const onChange = vi.fn();
  const decoratedStreams = decorateAllStreams(streams);

  beforeEach(() => {
    allStreams.value = decoratedStreams;
    loaded.value = true;
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("matches snapshot", () => {
    const { asFragment, getBySelector } = renderWithProviders(
      <AllStreamsContext.Provider value={useAllStreamsHookMock}>
        <StreamSelection
          name="sel"
          required
          value={decoratedStreams[0]}
          onChange={onChange}
        />
      </AllStreamsContext.Provider>
    );
    expect(getBySelector('select').className).toEqual('form-select is-valid');
    expect(asFragment()).toMatchSnapshot();
  });

  test('select with an error', () => {
    const { getBySelector } = renderWithProviders(
        <AllStreamsContext.Provider value={useAllStreamsHookMock}>
          <StreamSelection
            name="sel"
            required
            value={decoratedStreams[0]}
            onChange={onChange}
            error="value is bad"
          />
        </AllStreamsContext.Provider>
      );
      expect(getBySelector('select').className).toEqual('form-select is-invalid');
  });

  test('select with no value', () => {
    const { getBySelector } = renderWithProviders(
        <AllStreamsContext.Provider value={useAllStreamsHookMock}>
          <StreamSelection
            name="sel"
            required
            onChange={onChange}
          />
        </AllStreamsContext.Provider>
      );
      expect(getBySelector('select').className).toEqual('form-select');
  });

  test('change value', async () => {
    const user = userEvent.setup();
    const { getBySelector } = renderWithProviders(
        <AllStreamsContext.Provider value={useAllStreamsHookMock}>
          <StreamSelection
            name="sel"
            required
            value={decoratedStreams[0]}
            onChange={onChange}
          />
        </AllStreamsContext.Provider>
      );
      const elt = getBySelector('select') as HTMLSelectElement;
      await user.selectOptions(elt, [decoratedStreams[1].title]);
      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith({ name: "sel", value: decoratedStreams[1].pk});
  });
});
