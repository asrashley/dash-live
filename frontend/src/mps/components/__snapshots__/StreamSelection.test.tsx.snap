// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StreamSelection component > matches snapshot 1`] = `
<DocumentFragment>
  <select
    class="form-select is-valid"
    name="sel"
    required=""
  >
    <option
      value=""
    >
      --Select a stream--
    </option>
    <option
      value="1"
    >
      Big Buck Bunny
    </option>
    <option
      value="2"
    >
      Tears of Steel
    </option>
  </select>
</DocumentFragment>
`;
